src/js/components/AppListComponent.jsx

```diff
@@ -271,60 +314,61 @@ var AppListComponent = React.createClass({
       });
     }
 
-    return nodesSequence;
+    return items;
   },
 
-  getGroupedNodes: function (apps, filterCounts) {
-    var currentGroup = this.props.currentGroup;
-
-    var appsInGroup = apps.filter(app => app.id.startsWith(currentGroup));
+  groupItems: function (items, currentGroup, filterCounts) {
+    //  Exclude items that are not in the current group
+    var itemsInCurrentGroup = items.filter(
+      item => item.id.startsWith(currentGroup)
+    );
 
-    appsInGroup.forEach(app => {
-      filterCounts.appsStatusesCount[app.status]++;
-      if (app.container != null && app.container.volumes != null) {
+    itemsInCurrentGroup.forEach(item => {
+      if (item.isGroup) {
+        return;
+      }
+      filterCounts.appsStatusesCount[item.status]++;
+      if (item.container != null && item.container.volumes != null) {
         filterCounts.appsVolumesCount++;
       }
-      app.health.forEach(health => {
+      item.health.forEach(health => {
         if (health.quantity) {
           filterCounts.appsHealthCount[health.state]++;
         }
       });
     });
 
-    return appsInGroup
-      .reduce((memo, app) => {
-        let relativePath = app.id.substring(currentGroup.length);
-        let pathParts = relativePath.split("/");
-        let isGroup = pathParts.length > 1;
+    // Exclude items that are in a sub group
+    itemsInCurrentGroup = itemsInCurrentGroup.filter(item => {
+      let relativePath = item.id.substring(currentGroup.length);
+      let pathParts = relativePath.split("/");
+      return pathParts.length < 2;
+    });
 
-        if (!isGroup) {
-          memo.push(app);
-        } else {
-          let groupId = currentGroup + pathParts[0];
-          let group = memo.find(item => {
-            return item.id === groupId;
-          });
-
-          if (group == null) {
-            group = initGroupNode(groupId, app);
-            memo.push(group);
-          } else {
-            updateGroupNode(group, app);
-          }
-        }
-        return memo;
-      }, []);
+    return itemsInCurrentGroup;
```

Comment:
We iterate over `items` and subsets of `items` repeatedly here. Not to get too interested in optimisation, but shouldn't we try to combine at least the last two items?